### PR TITLE
feat-logger-update

### LINKED ISSUE #1056: bug: Multiprocess cannot write log to the same file
**Problem**
In windows, once a file is taken by one process, other process can not modify it. For example, when the server is writing to log file, it hold `cortex.log`, and if we run CLI command, it cannot write message to `cortex.log`.

**Solution**
In window, we are planning to use Unix Domain Socket to start a thread and write log to file. Below are detail implementation:
- When start server, will create a thread to run UDS to listen messages and write to logfile.
- From server side, override log method of Trantor to send log message to UDS server
- From CLI side, when execute command, check if UDS server is live, if not write directly log to log file.
- With command has `--verbose` flag, log everything to terminal.

### LINKED ISSUE #1153: epic: Cortex.cpp logs
## Goal

- Strong logging infra for Server and CLI
- Users can find and submit logfiles easily in case of bugs
- Logs should ideally not have PII for privacy (e.g. if published on Github)

## Open Questions

- [ ] Should we have separate logs for Server and CLI? (remove needs for UDS)
- [ ] What is our log pruning strategy? 

## Tasklist

- [x] #1054 
- [x] https://github.com/janhq/cortex.cpp/issues/1038
- [x] https://github.com/janhq/cortex.cpp/issues/1007
- [x] https://github.com/janhq/cortex.cpp/issues/1062

### PR DESCRIPTION
**Pull request by nguyenhoangthuan99**
Fixes #1153

### CODE REVIEW

**Comment by nguyenhoangthuan99** [comment]
- [x] separate log file for `cortex.log` and `cortex-cli.log`
- [x] pruning log file only to save last 100,000 lines. The log file will be truncated if number of  line > max_lines. Check if file will be truncated every 1000 new lines.
- [x] configurable log file location. Save the config in the ~/.cortexrc 
```
logFolderPath: C:\Users\jan\cortexcpp
dataFolderPath: C:\Users\jan\cortexcpp
host: 127.0.0.1
port: 3928
```

**Comment by dan-menlo** [comment]
@nguyenhoangthuan99 Please link this PR to issue via keyword - I've helped to do this in the Description

**Comment by dan-menlo** [comment]
@nguyenhoangthuan99 A few questions: 

- Do we have a filesize estimate of how much space 100,000 log lines will equate to? 
- Should we make num log lines a variable in `.cortexrc` that users can configure? (need to find a good variable name for this)

I imagine cortex.cpp may have some embedded use-cases in the future, where devs may have diskspace constraints for logs (difficult for me to imagine now with >1gb model sizes, but this may change in the future)

File: engine/utils/cortex_utils.h [hunk 1]
```diff
@@ -34,8 +34,10 @@ constexpr static auto kTensorrtLlmPath = "/engines/cortex.tensorrt-llm";
 
 inline std::string models_folder = "./models";
 inline std::string logs_folder = "./logs";
-inline std::string logs_base_name = "./logs/cortex";
+inline std::string logs_base_name = "./logs/cortex.log";
+inline std::string logs_cli_base_name = "./logs/cortex-cli.log";
 inline size_t log_file_size_limit = 20000000;  // ~20 mb
+inline size_t log_file_max_lines = 100000;     // 100,000 lines
```
Comment (dan-menlo) [inline engine/utils/cortex_utils.h]: @nguyenhoangthuan99 I think we should have this as a variable from `.cortexrc`

**Review by dan-menlo — CHANGES_REQUESTED** [review]

File: engine/utils/cortex_utils.h [hunk 2]
```diff
@@ -34,8 +34,10 @@ constexpr static auto kTensorrtLlmPath = "/engines/cortex.tensorrt-llm";
 
 inline std::string models_folder = "./models";
 inline std::string logs_folder = "./logs";
-inline std::string logs_base_name = "./logs/cortex";
+inline std::string logs_base_name = "./logs/cortex.log";
+inline std::string logs_cli_base_name = "./logs/cortex-cli.log";
 inline size_t log_file_size_limit = 20000000;  // ~20 mb
```
Comment (dan-menlo) [inline engine/utils/cortex_utils.h]: @nguyenhoangthuan99 Is this variable still used?

**Review by dan-menlo — CHANGES_REQUESTED** [review]

File: engine/utils/cortex_utils.h (hunk 2 above)
Comment (nguyenhoangthuan99) [reply engine/utils/cortex_utils.h]: for now, we don't use the file size limit anymore. I will update it

File: engine/utils/cortex_utils.h (hunk 1 above)
Comment (nguyenhoangthuan99) [reply engine/utils/cortex_utils.h]: I'll move `log_file_max_lines` to read from .cortexrc

**Comment by nguyenhoangthuan99** [comment]
Hi @dan-homebrew , Average each line will contains < 300 characters , each character is 1 byte. So 100,000 lines will be 300 * 100,000/(1024*1024) ~ 30Mb

**Comment by dan-menlo** [comment]
> Hi @dan-homebrew , Average each line will contains < 300 characters , each character is 1 byte. So 100,000 lines will be 300 * 100,000/(1024*1024) ~ 30Mb

This looks good to me - let's include it as a default param in `.cortexrc`

File: engine/main.cc [hunk 3]
```diff
@@ -29,16 +30,18 @@ void RunServer() {
   LOG_INFO << "Host: " << config.apiServerHost << " Port: " << config.apiServerPort << "\n";
 
   // Create logs/ folder and setup log to file
-  std::filesystem::create_directory(cortex_utils::logs_folder);
-  trantor::AsyncFileLogger asyncFileLogger;
-  asyncFileLogger.setFileName(cortex_utils::logs_base_name);
+  std::filesystem::create_directory(config.logFolderPath + "/" +
+                                    cortex_utils::logs_folder);
+  trantor::FileLogger asyncFileLogger;
+  asyncFileLogger.setFileName(config.logFolderPath + "/" +
```
Comment (namchuai) [inline engine/main.cc:36]: not sure if `/` works in windows. If you tested and it's work then please ignore this comment.

Comment (nguyenhoangthuan99) [reply engine/main.cc:36]: I tested in windows and it can works._

File: engine/utils/file_logger.cc [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+  OpenFile();
+  LoadExistingLines();
+  TruncateFileIfNeeded();
+}
+
+FileLogger::CircularLogFile::~CircularLogFile() {
+  CloseFile();
+}
+void FileLogger::CircularLogFile::writeLog(const char* logLine,
+                                           const uint64_t len) {
```
Comment (nguyenhoangthuan99) [inline engine/utils/file_logger.cc:42]: need to add mutex here for thread safe

File: engine/utils/config_yaml_utils.h [hunk 5]
```diff
@@ -48,8 +53,16 @@ inline CortexConfig FromYaml(const std::string& path,
 
   try {
     auto node = YAML::LoadFile(config_file_path.string());
+    int max_lines_;
```
Comment (vansangpfiev) [inline engine/utils/config_yaml_utils.h]: naming convention: should rename to `max_lines`

File: engine/utils/file_logger.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <unistd.h>
+#endif
+
+namespace trantor {
+
+class TRANTOR_EXPORT FileLogger : public AsyncFileLogger {
```
Comment (vansangpfiev) [inline engine/utils/file_logger.h:19]: Please mark function with `override` keyword if implements virtual function inherit from `AsyncFileLogger`

**Review by dan-menlo — APPROVED** [review]
Lgtm

File: engine/utils/file_logger.h [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    ~CircularLogFile();
+
+    void writeLog(const char* logLine, const uint64_t len);
+    void flush();
+    uint64_t getLength() const { return totalLines_; }
+
+   private:
+    FILE* fp_{nullptr};
+    uint64_t max_lines_;
+    uint64_t totalLines_{0};
```
Comment (vansangpfiev) [inline engine/utils/file_logger.h]: I'm not sure why we need `totalLines_` variable. Can you please elaborate?

Comment (nguyenhoangthuan99) [reply engine/utils/file_logger.h]: Seem like we don't need it, I removed `totalLines_`

**Review by vansangpfiev — APPROVED** [review]

**Comment by dan-menlo** [comment]
Great job!

**Pull request merged**